File: 5.双指针法/9.三数之和/15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < nums.size(); i ++ )
        {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            int target = - nums[i];
            int l = i + 1, r = nums.size() - 1;
            while (l < r)
            {
                if (nums[l] + nums[r] == target)
                {
                    res.push_back({nums[i], nums[l], nums[r]});
                    while (l < r && nums[r] == nums[r - 1]) r --;
                    while (l < r && nums[l] == nums[l + 1]) l ++;
                    l ++, r --;
                }
                else if (nums[l] + nums[r] > target)
                {
                    r --;
                }
                else l ++;
            }
        }
        return res;
    }
};
// @lc code=end

File: 面经150/面经150p1/2.双指针/15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#
from mytools import *
# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        n = len(nums)
        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]: continue
            tar = -nums[i]
            l = i + 1; r = n - 1
            while l < r:
                if nums[l] + nums[r] == tar:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1;r -= 1
                    while l < r and nums[l] == nums[l - 1]: l += 1
                    while l < r and nums[r] == nums[r + 1]: r -= 1
                elif nums[l] + nums[r] > tar:
                    r -= 1
                    while l < r and nums[r] == nums[r + 1]: r -= 1
                else:
                    l += 1
                    while l < r and nums[l] == nums[l - 1]: l += 1
        return res
# @lc code=end

